src/mlpack/core/optimizers/cne/cne_impl.hpp

```diff
@@ -0,0 +1,227 @@
+{
+  // Make sure for evolution to work at least four candidates are present.
+  if (populationSize < 4)
+    throw std::logic_error("Population size should be atleast 4!");
+
+  // Find the number of elite canditates from population.
+  numElite = floor(selectPercent * populationSize);
+
+  // Making sure we have even number of candidates to remove and create.
+  if ((populationSize - numElite) % 2 != 0) numElite--;
+
+  // Terminate if two parents can not be created.
+  if (numElite < 2)
+    throw std::logic_error("Increase selection percentage.");
+
+  // Terminate if at least two childs are not possible.
+  if ((populationSize - numElite) < 2 )
+    throw std::logic_error("Increase population size.");
+
+  // Get the number of functions to iterate.
+  const size_t numFun = function.NumFunctions();
+
+  // Set the population size and fill random values [0,1].
+  population = arma::randu(iterate.n_rows, iterate.n_cols, populationSize);
+
+  // Store the number of elements in a cube slice or a matrix column.
+  elements = population.n_rows * population.n_cols;
+
+  // initializing helper variables.
+  fitnessValues.set_size(populationSize);
+  double fitness = 0;
+  double lastBestFitness = 0;
+
+  Log::Info << "CNE initialized successfully. Optimization started"
+      << std::endl;
+
+  // Find the fitness before optimization using given iterate parameters.
+  for (size_t j = 0; j < numFun; j++)
+    lastBestFitness += function.Evaluate(iterate, j);
+
+  // Iterate until maximum number of generations is obtained.
+  for (size_t gen = 1; gen <= maxGenerations; gen++)
+  {
+    // Calculating fitness values of all candidates.
+    for (size_t i = 0; i < populationSize; i++)
+    {
+       // Select a candidate and insert the parameters in the function.
+       iterate = population.slice(i);
+
+       // Find fitness of candidate.
+       for (size_t j = 0; j < numFun; j++)
+         fitness += function.Evaluate(iterate, j);
+
+       // Save fitness value of the evaluated candidate.
+       fitnessValues[i] = fitness;
+       fitness = 0;
+    }
+
+      Log::Info << "Generation number: " << gen << " best fitness = "
+          << fitnessValues.min() << std::endl;
+
+      // Create next generation of species.
+      Reproduce();
+
+      // Check for termination criteria.
+      if (tolerance >= fitnessValues.min())
+      {
+          Log::Info << "Terminating. Given fitness criteria " << tolerance
+              << " > " << fitnessValues.min() << std::endl;
+          break;
+      }
+
+      // Check for termination criteria.
+      if (lastBestFitness - fitnessValues.min() < objectiveChange)
+      {
+          Log::Info << "Terminating. Fitness History change "
+              << (lastBestFitness - fitnessValues.min())
+              << " < " << objectiveChange << std::endl;
+          break;
+      }
+
+      // Store the best fitness of present generation.
+      lastBestFitness = fitnessValues.min();
```

Comment:
This block (starting with the `Generation number:` output) seems to be indented two spaces too many.